### ISSUE TITLE
[feature request] More samplers.

### ISSUE THREAD

**Issue by Rishav1** [pfnet/optuna#346]
It would be amazing if support for more samplers and pruners are added. Some Meta-heuristic search algorithms like simulated annealing or hill climbing can be easily implemented. However, more complex methods like population based techniques (evolutionary algorithms, particle swarm optimization) could be challenging to implement given your current schema. 

It would be nice to have more samplers in my opinion.

**Comment by sile** [comment]
Thank you for your suggestion.

Optuna dev team is also planning to make more sampling/pruning algorithms available.
As part of that, we are planning to expose the interface of [sampler] and [pruner] so that users can implement their own samplers and pruners.
Now, we are working on interface design, and it will be delivered in the near future.

About complex algorithms, as you say, it is not easy to develop them within the current Optuna framework.
We are considering how to integrate those algorithms with Optuna, but it may take some time before we can provide them.

[sampler]: https://github.com/pfnet/optuna/blob/v0.8.0/optuna/samplers/base.py#L9
[pruner]: https://github.com/pfnet/optuna/blob/v0.8.0/optuna/pruners/base.py#L8

**Comment by g-votte** [comment]
Closing, since we're going to have more samplers such as #399
